### PR TITLE
Integrate RandomX pow

### PR DESCRIPTION
**Pull request by poszu**
Integration of [RandomX](https://github.com/tevador/RandomX)-based proof of work in place of the scrypt-based one.

### Changes
- PoW difficulty is now 256bit, configured in form of `[32]byte` array;
- verifier is now stateful because RandomX cache takes about 250ms to initialize, which is too long to do it per ATX;
- proving and verifying options take `PowFlags` that configure the operation of RandomX. They have sane defaults:
  - proving runs in FAST mode (2GiB RAM)
  - verification runs in LIGHT mode (256MiB RAM)
- renamed "K2 PoW" to just "PoW" everywhere. Reasoning - there is only one PoW in POST now AND the old name was uninformative anyway.

:bulb: configuration of RandomX flags affect speed and memory requirements only - misconfiguring it won't affect the generated pow nonce correctness.

:bulb: `postrs.Verifier` can be used from many threads safely.

### CODE REVIEW

File: internal/postrs/proof.go [hunk 1]
```diff
@@ -44,10 +46,11 @@ func GenerateProof(dataDir string, challenge []byte, cfg config.Config, logger *
...
-		k1:                C.uint32_t(cfg.K1),
-		k2:                C.uint32_t(cfg.K2),
-		k2_pow_difficulty: C.uint64_t(cfg.K2PowDifficulty),
-		pow_scrypt:        translateScryptParams(powScrypt),
+		k1: C.uint32_t(K1),
+		k2: C.uint32_t(K2),
+	}
+	for i, b := range powDifficulty {
+		config.pow_difficulty[i] = C.uchar(b)
 	}
```
Comment (fasmat) [inline internal/postrs/proof.go:54]: Why not do the same for the difficulty as for the challenge?

```golang
	diffPtr := C.CBytes(powDifficulty)
	defer C.free(diffPtr)

	config := C.Config{
		k1: C.uint32_t(K1),
		k2: C.uint32_t(K2),
		pow_difficulty: (*C.uchar)(diffPtr),
	}
```

avoids copying the data, but requires `pow_difficulty` to be a pointer instead of an array.

File: proving/proving.go [hunk 2]
```diff
@@ -16,9 +16,9 @@ import (
 
 func Generate(ctx context.Context, ch shared.Challenge, cfg config.Config, logger *zap.Logger, opts ...OptionFunc) (*shared.Proof, *shared.ProofMetadata, error) {
 	options := option{
-		threads:   1,
-		nonces:    16,
-		powScrypt: config.DefaultPowParams(),
+		threads:  1,
+		nonces:   16,
+		powFlags: postrs.GetRecommendedPowFlags() | postrs.PowFastMode,
```
Comment (fasmat) [inline proving/proving.go]:
```suggestion
		powFlags: config.DefaultProvingPowFlags(),
```

File: config/config.go [hunk 3]
```diff
@@ -45,19 +79,26 @@ type Config struct {
...
+		K1:              26,
+		K2:              37,
+		K3:              37,
 		K2PowDifficulty: 0x0FFFFFFF_FFFFFFFF,
 	}
+	for i := range cfg.PowDifficulty {
+		cfg.PowDifficulty[i] = 0xFF
+	}
+
+	return cfg
```
Comment (fasmat) [inline config/config.go:101]: With `PowDifficulty` as a slice this can be written more concisely as:

```golang
	return Config{
		LabelsPerUnit:   512, // 8kB units
		MaxNumUnits:     defaultMaxNumUnits,
		MinNumUnits:     defaultMinNumUnits,
		K1:              26,
		K2:              37,
		K3:              37,
		K2PowDifficulty: 0x0FFFFFFF_FFFFFFFF,
		PowDifficulty:   bytes.Repeat(0xFF, 32),
	}
}
```

The check for the length then has to be done in `GenerateProof` and `VerifyProof` (where we are already checking for other lengths as well).

It would also allow to pass the difficulty without the need to copy it first (see other comment).

File: internal/postrs/proof.go [hunk 4]
```diff
@@ -69,11 +73,78 @@ func GenerateProof(dataDir string, challenge []byte, cfg config.Config, logger *
...
+	if result != C.Ok {
+		return nil, fmt.Errorf("failed to create verifier")
+	}
+	return &verifier, nil
+}
+
+func (v *Verifier) Close() error {
+	v.closeOnce.Do(func() { C.free_verifier(v.inner) })
+	return nil
+}
```
Comment (fasmat) [inline internal/postrs/proof.go:132]: Should multiple calls to close return an error?

File: verifying/verifying.go [hunk 5]
```diff
@@ -59,9 +59,35 @@ func VerifyVRFNonce(nonce *uint64, m *shared.VRFNonceMetadata, opts ...OptionFun
 	return nil
 }
 
+type ProofVerifier struct {
+	inner *postrs.Verifier
+}
```
Comment (fasmat) [inline verifying/verifying.go]:
```suggestion
type ProofVerifier struct {
	*postrs.Verifier
}
```

With struct embedding `ProofVerifier` inherits the methods of `postrs.Verifier` (no need to define a `Close` method that just calls the inner `Close`)

File: verifying/verifying.go [hunk 6]
```diff
@@ -59,9 +59,35 @@ func VerifyVRFNonce(nonce *uint64, m *shared.VRFNonceMetadata, opts ...OptionFun
...
+
+// NewProofVerifier creates a new proof verifier.
+// If powFlags is nil, the recommended PoW flags will be used.
+// The verifier must be closed after use with Close().
+func NewProofVerifier(powFlags *config.PowFlags) (*ProofVerifier, error) {
```
Comment (fasmat) [inline verifying/verifying.go]: I think we should keep the functional options for optional parameters and not use `nil` as a placeholder for "default parameter".

We are already using them in the `proving` package and adding them here in the `verifying` package would make a more consistent API.

File: internal/postrs/proof.go (hunk 4 above)
Comment (poszu) [reply internal/postrs/proof.go:132]: No, I wanted it to be idempotent

File: internal/postrs/proof.go (hunk 1 above)
Comment (poszu) [reply internal/postrs/proof.go:54]: It doesn't avoid copying the data because `C.CBytes(powDifficulty)` allocates memory and makes a copy. It will only add a layer of indirection and lose the benefit of having the length embedded in the type (`[32]byte` vs `[]byte`), which makes the API easier to use properly.

File: config/config.go (hunk 3 above)
Comment (poszu) [reply config/config.go:101]: I prefer to put the required length into the type. Then, the length is obvious and manual validation is not required.

File: internal/postrs/proof.go (hunk 1 above)
Comment (fasmat) [reply internal/postrs/proof.go:54]: `[32]byte` is a value type so it is copied when `C.generate_proof` function is called a bit further down, while `C.CBytes` returns a pointer to the data that doesn't need to be copied again.

I don't think it makes much difference between arrays and slices in terms of what is easier to use. We even use `[]byte` over `[x]byte` for most other types here, passing fewer bytes to the function also just fils them with zeros instead of causing an error during compilation: https://go.dev/play/p/i_5jDu8yc50

Comment (poszu) [reply internal/postrs/proof.go:54]: >[32]byte is a value type so it is copied when C.generate_proof function is called a bit further down, while C.CBytes returns a pointer to the data that doesn't need to be copied again.

That's the same amount of copying, except `C.CBytes()` allocates on a heap (and copies the contents into there) while the array is kept on a stack. Anyhow, it's supposed to be called once every epoch.

>I don't think it makes much difference between arrays and slices in terms of what is easier to use.

I mean that an array, which is fixed length better expresses the API requirements.
- `[32]byte` says: "Pass me **exactly 32** bytes." 
- `[]byte` says: "Pass me **some** bytes

In other words, when I see a function signature:
```go
func Foo(val [32]byte)
```
- I immediately know that I must pass 32 bytes,
- I don't need to validate length of passed data (which might be necessary if `[]byte` is passed).

Comment (fasmat) [reply internal/postrs/proof.go:54]: > That's the same amount of copying, except C.CBytes() allocates on a heap (and copies the contents into there) while the array is kept on a stack.

Array does 2 copies: here in this `for` loop and again when calling `C.generate_proof` by passing the array via value to the function.

Slice only once: when calling `C.CBytes`, after that only the pointer to the array is passed to `C.generate_proof`.

Comment (fasmat) [reply internal/postrs/proof.go:54]: But you are right it's not a necessary optimization because generating a proof is only done every 2 weeks and compared to the rest of the process its an negligible improvement.

**Review by fasmat — APPROVED** [review]
